### PR TITLE
Isolate Cereal

### PR DESCRIPTION
**Pull request by akleeman**
**It should be helpful to review each commit in order.**

This change was motivated by discovering that 1) `cereal` headers are pretty heavy 2) using member serialization methods required including cereal in just about any `albatross` include. 3) the use of polymorphic registration generates a TON of symbols (alluded to in the "Registering in a header file" section here: https://uscilab.github.io/cereal/polymorphism.html) which slows down linking.

The change is effectively two parts, in the first we remove polymorphism from `priors.hpp` which was the only part of `albatross` that was actually using polymorphic pointers.  This eliminates the need for making `cereal` work with polymorphism via the `CEREAL_REGISTER_TYPE` which blew up linking (it added ~10k symbols to our static libraries!!).  Second all serialization methods were moved from member methods to free functions in a separate directory.  The result is the ability to include albatross without including cereal, then only include `albatross/Serialization` in `.cc` files which require serialization.

NOTE: The result cut the unit test time in half!

### CODE REVIEW

**Comment by akleeman** [comment]
I ended up using a strange pattern which *seems* to work quite nicely for this application.  @jbangelo I'm curious if you have thoughts.  Basically instead of using polymorphic pointers I switched to using variants which I then wrapped in a container.   Ie, given a situation with

```
struct Base {
  int bar() const = 0;
}

struct X : Base {
  int bar() const override {return 1;};
}
```
instead of passing everything around as a `std::shared_ptr<Base>` I instead created a container:
```
struct PolymorphicContainer : Prior {
  int bar() const override {
    return foo.match([](const auto &f) {return f.bar()};
  }  
  variant<X, Y, Z> foo;
}
```
which also inherits from `Base` but implements all virtual methods using `variant<>::match`.  The inheritance pattern seems compatible with the container and results in much more intuitive compiler errors if one of the types in the variant doesn't support a method it should. 

The upside for this application is that it means I don't need to deal with registering polymorphic types and the mess that results from serializing pointers in `cereal`, this now happens with the serialization of the `variant`. The downside of this is that you need to know all the available types when the container is declared, so someone using `albatross` would not be able to add custom priors (though perhaps there's a work around for that).

make sense?

File: tests/test_serialize.cc [hunk 1]
```diff
@@ -266,7 +288,7 @@ typedef ::testing::Types<LDLT, ExplainedCovarianceRepresentation, EigenMatrix3d,
                          EigenVectorXd, EmptyEigenMatrixXd, EigenMatrixXd,
                          FullJointDistribution, MeanOnlyJointDistribution,
                          FullMarginalDistribution, MeanOnlyMarginalDistribution,
-                         ParameterStoreType, Dataset, DatasetWithMetadata,
+                         Dataset, DatasetWithMetadata, ParameterStoreType,
```
[the remark this answers is not in the document]
Comment (akleeman) [reply tests/test_serialize.cc]: Ah, that was just left over from when I was commenting out and rearranging the order to debug some compile errors.  I'll revert it to avoid confusion.

File: include/albatross/src/cereal/ransac.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+template <typename Archive, typename InlierMetric, typename IndexingFunction>
+inline void serialize(
+    Archive &archive,
+    GaussianProcessRansacStrategy<InlierMetric, IndexingFunction> &strategy,
+    const std::uint32_t) {
+  archive(cereal::make_nvp("inlier_metric", strategy.inlier_metric_));
+  archive(cereal::make_nvp("indexing_function", strategy.indexing_function_));
+}
+
+// template <typename Archive, typename ModelType, typename StrategyType>
```
[the remark this answers is not in the document]
Comment (akleeman) [reply include/albatross/src/cereal/ransac.hpp]: thanks!

File: include/albatross/src/cereal/parameters.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+namespace cereal {
+
+template <class Archive>
+inline void serialize(Archive &archive, albatross::Parameter &param,
+                      const std::uint32_t) {
+  archive(cereal::make_nvp("value", param.value));
+  archive(cereal::make_nvp("prior", param.prior));
+};
+
+// template <class Archive>
```
[the remark this answers is not in the document]
Comment (akleeman) [reply include/albatross/src/cereal/parameters.hpp]: thanks!

File: tests/test_indexing.cc [hunk 4]
```diff
@@ -124,7 +125,7 @@ TEST(test_indexing, test_diagonal_set_subset) {
   expected = expected_diag.asDiagonal();
   to = DiagonalMatrixXd(x);
   set_subset(from, idx, &to);
-  EXPECT_EQ(to, expected);
+  EXPECT_TRUE(to == expected);
```
Comment (jbangelo) [inline tests/test_indexing.cc]: Why was this needed? Doesn't `EXPECT_EQ` use `==`?

File: include/albatross/serialize/Common [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+#include <cereal/cereal.hpp>
+#include <cereal/archives/binary.hpp>
+#include <cereal/archives/json.hpp>
+#include <cereal/archives/xml.hpp>
+#include <cereal/types/map.hpp>
+#include <cereal/types/string.hpp>
+#include <cereal/types/vector.hpp>
+
+#include "../src/cereal/traits.hpp"
```
Comment (jbangelo) [inline include/albatross/serialize/Common]: Relative paths like this in include paths aren't great (they tend to break when porting to different systems(, and cause generally be avoided by setting the include paths you want. Would a path like `albatross/src/cereal/traits.hpp` work?

File: tests/test_indexing.cc (hunk 4 above)
Comment (akleeman) [reply tests/test_indexing.cc]: I have no clue why `EXPECT_EQ` doesn't work with `Eigen::DiagonalMatrix<double, Dynamic>`.  It probably has something to do with the lack of `==` operator (I had to define a free standing one).

File: include/albatross/Serialization [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#include "src/cereal/evaluation.hpp"
+#include "src/cereal/ransac.hpp"
+#include "src/cereal/gp.hpp"
+#include "src/cereal/least_squares.hpp"
+#include "src/cereal/model.hpp"
+#include "src/cereal/parameters.hpp"
+#include "src/cereal/representations.hpp"
+#include "src/cereal/variant.hpp"
+
+//CEREAL_SPECIALIZE_FOR_ALL_ARCHIVES(albatross::ParameterStore, cereal::specialization::non_member_load_save );
```
[the remark this answers is not in the document]
Comment (akleeman) [reply include/albatross/Serialization]: Nope!  thanks.
